picky/src/ssh/certificate.rs

```diff
@@ -0,0 +1,854 @@
+
+    /// Optional
+    pub fn extensions(&self, extensions: Vec<SshExtension>) -> &Self {
+        self.inner.borrow_mut().extensions = Some(extensions);
+        self
+    }
+
+    /// Required
+    pub fn signature_key(&self, signature_key: &'a SshPrivateKey) -> &Self {
+        self.inner.borrow_mut().signature_key = Some(signature_key);
+        self
+    }
+
+    /// Optional. RsaPkcs1v15 with SHA256 is used by default.
+    pub fn signature_algo(&self, signature_algo: SignatureAlgorithm) -> &Self {
+        self.inner.borrow_mut().signature_algo = Some(signature_algo);
+        self
+    }
+
+    /// Optional
+    pub fn comment(&self, comment: String) -> &Self {
+        self.inner.borrow_mut().comment = Some(comment);
+        self
+    }
+
+    pub fn build(&self) -> Result<SshCertificate, SshCertificateGenerationError> {
+        let mut inner = self.inner.borrow_mut();
+
+        let SshCertificateBuilderInner {
+            cert_key_type,
+            public_key,
+            serial,
+            cert_type,
+            key_id,
+            valid_principals,
+            valid_after,
+            valid_before,
+            critical_options,
+            extensions,
+            signature_algo,
+            signature_key,
+            comment,
+        } = inner.deref_mut();
+
+        let cert_key_type = cert_key_type.ok_or(SshCertificateGenerationError::NoKeyType)?;
+        match cert_key_type {
+            SshCertKeyType::SshRsaV01 | SshCertKeyType::RsaSha2_256V01 | SshCertKeyType::RsaSha2_512v01 => {}
+            SshCertKeyType::EcdsaSha2Nistp256V01
+            | SshCertKeyType::SshDssV01
+            | SshCertKeyType::EcdsaSha2Nistp384V01
+            | SshCertKeyType::EcdsaSha2Nistp521V01
+            | SshCertKeyType::SshEd25519V01 => {
+                return Err(SshCertificateGenerationError::UnsupportedCertificateKeyType(
+                    cert_key_type.as_str().to_owned(),
+                ))
+            }
+        }
+
+        let public_key = public_key
+            .take()
+            .ok_or(SshCertificateGenerationError::MissingPublicKey)?;
+        let serial = serial.take().unwrap_or(0);
+        let cert_type = cert_type
+            .take()
+            .ok_or(SshCertificateGenerationError::MissingCertificateType)?;
+        let key_id = key_id.take().unwrap_or_default();
+
+        let mut nonce = Vec::new();
+        let mut rnd = rand::thread_rng();
+        for _ in 0..32 {
+            nonce.push(rnd.gen::<u8>());
+        }
+
+        let now_timestamp = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64;
+        let valid_after = valid_after.take().ok_or(SshCertificateGenerationError::InvalidTime)?;
+        let valid_before = valid_before.take().ok_or(SshCertificateGenerationError::InvalidTime)?;
+        if valid_after.timestamp() > now_timestamp || now_timestamp >= valid_before.timestamp() {
+            return Err(SshCertificateGenerationError::InvalidTime);
+        }
```

Comment:
I missed that in previous review: now's time shouldn't be retrieved and checked against using stuff like `SystemTime::now()` or `OffsetDateTime::now_utc()`. You may validate that `valid_after < valid_before` though. As long as this property holds, there is no reason to reject the certificate creation.